### DISCUSSION TITLE
Is it possible to use both /api and /pages/api functions at the same time?

### DISCUSSION THREAD

**Discussion by jesper-bylund** [vercel/next.js#37634] [Help]
For obvious reasons this is hard to search for.

Currently calls to `/pages/api` are sorting query params on the call, which can cause errors (in my case, calls from Hubspot API are not possible to verify since the params have been changed since the hash was created).

It seems like you can reach the raw request with `/api` functions though. So in my case I would like to use functions in both `/api` and `/pages/api` at the same time. 
Is it possible, and if so, how?

**Comment by icyJoseph** [comment]
~~I think if you are deploying to Vercel, you can use the Vercel feature that makes it so that top level `api` folders are also lambda functions. For example if you deploy a CRA, with a top level `api` folder, Vercel uses that as entry point for lambdas. [Reference](https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions).~~

~~AFAIK, `api` has priority over `pages/api`, meaning that `api/hello` makes `pages/api/hello` unreachable. This would mean switching from the Next.js CLI to the Vercel CLI, when working locally. Pretty much making you absolutely dependant on their solution, and not very elegant I'd say, but since you ask, I think it should be possible.~~

### Alternative

For your specific case, perhaps try to export this config from our `pages/api` file:

```js
export const config = {
  api: {
    bodyParser: false,
  },
}
```

Then you can work the incoming request raw, and do as you wish. More info [here](https://nextjs.org/docs/api-routes/api-middlewares#custom-config).

**Reply by leerob** [reply]
Yeah, you can use the raw body in both Next.js API Routes as well and general APIs on Vercel: https://vercel.com/support/articles/how-do-i-get-the-raw-body-of-a-serverless-function

**Reply by leerob** [reply]
If you're using Next.js on Vercel, you shouldn't need to use `/api` at all, since it's the same thing as `pages/api`.

**Comment by jesper-bylund** [comment]
Hi @leerob,
this is not about the body. This is about the query params. The actual URI is being reschuffled in `pages/api`. Which is why I want to use `/api` where this problem shouldn't be present (according to a forum post I found, haven't tried it yet).

**Reply by leerob** [reply]
Could you share a specific code example of what you're trying to do and what is not working?

**Reply by jesper-bylund** [reply]
I'm trying to get the requested URI. I don't know if you need code for that?

It's used in this signature:
https://developers.hubspot.com/changelog/introducing-version-3-of-webhook-signatures

When using the logs on vercel you can see the called URI (and in hubspots monitoring tool as well ofc). These match perfectly. But in the functions `req.url` has re-sorted the query parameters.

This:
`https://test.vercel.app/hook?test1=first&test2=second`
becomes this:
`https://test.vercel.app/hook?test2=second&test1=first`

That makes signatures impossible.

**Reply by jesper-bylund** [reply]
To be very clear, my code runs perfectly if I copy the URI from the vercel log into my function. That works. But `req.url` does not. Since it's different.
